File: problem350.cs
public class Solution {
    public int[] Intersect(int[] nums1, int[] nums2) {
        Dictionary<int, int> count1 = new Dictionary<int, int>();
        Dictionary<int, int> count2 = new Dictionary<int, int>();

        foreach (int num in nums1) {
            if (!count1.ContainsKey(num)) {
                count1.Add(num, 0);
            }
            count1[num]++;
        }

        foreach (int num in nums2) {
            if (!count2.ContainsKey(num)) {
                count2.Add(num, 0);
            }
            count2[num]++;
        }

        IList<int> intersection = new List<int>();
        foreach (var kvp in count1) {
            if (count2.ContainsKey(kvp.Key)) {
                int count = (count1[kvp.Key] > count2[kvp.Key])
                    ? count2[kvp.Key]
                    : count1[kvp.Key];
                for (int j = 0; j < count; j++) {
                    intersection.Add(kvp.Key);
                }
            }
        }
        return intersection.ToArray();
    }
}

File: su19/problem350.py
class Solution(object):
    def intersect(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        if len(nums1) < len(nums2):
            smaller = nums1
            larger = nums2
        else:
            smaller = nums2
            larger = nums1
        counts = {}
        for elem in smaller:
            if elem not in counts:
                counts[elem] = 0
            counts[elem] += 1
        intersection = []
        for elem in larger:
            if elem in counts and counts[elem] > 0:
                intersection.append(elem)
                counts[elem] -= 1
        return intersection